File: python/data_structures/linked_lists/list_node.py
"""
Node class for singly linked list
"""

from typing import Optional


class ListNode:
    """A node in a singly linked list."""

    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        """
        Initialize a list node.

        Args:
            val: The value stored in the node
            next: Reference to the next node (None if this is the last node)
        """
        self.val = val
        self.next = next

File: rust/data-structures/src/linked_lists/list_node.rs
/// Node structure for singly linked list

#[derive(Debug, PartialEq, Eq)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    /// Create a new node with the given value
    pub fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }

    /// Create a new node with the given value and next pointer
    pub fn with_next(val: i32, next: Option<Box<ListNode>>) -> Self {
        ListNode { val, next }
    }
}
